php/TypeScript/PhpLexerBase.ts

```diff
@@ -0,0 +1,161 @@
+    protected _scriptTag: boolean;
+    protected _styleTag: boolean;
+    private _heredocIdentifier: string | undefined;
+    private _prevTokenType: number;
+    private _htmlNameText: string | undefined;
+    private _phpScript: boolean;
+    private _insideString: boolean;
+    protected _mode: number
+    protected _channel: number
+
+    static DEFAULT_MODE = 0;
+    protected static MORE = -2;
+    protected static SKIP = -3;
+
+    protected static DEFAULT_TOKEN_CHANNEL: number = (Token as any).DEFAULT_CHANNEL;
+    protected static HIDDEN: number = (Token as any).HIDDEN_CHANNEL;
+    protected static MIN_CHAR_VALUE = 0x0000;
+    protected static MAX_CHAR_VALUE = 0x10FFFF;
+
+    constructor(input: CharStream) {
+        super(input);
+        this.AspTags = true;
+        this._scriptTag = false;
+        this._styleTag = false;
+        this._heredocIdentifier = undefined;
+        this._prevTokenType = 0;
+        this._htmlNameText = undefined;
+        this._phpScript = false;
+        this._insideString = false;
+    }
+
+    nextToken() {
+        let token = super.nextToken()
+
+        if (token.type === PhpParser.PHPEnd || token.type === PhpLexer.PHPEndSingleLineComment) {
+            if (this._mode === PhpLexer.SingleLineCommentMode) {
+                // SingleLineCommentMode for such allowed syntax:
+                // // <?php echo "Hello world"; // comment ?>
+                this.popMode();
+            }
+            this.popMode();
+
+            if (token.text === "</script>") {
+                this._phpScript = false;
+                token.type = PhpLexer.HtmlScriptClose;
+            } else {
+                // Add semicolon to the end of statement if it is absent.
+                // For example: <?php echo "Hello world" ?>
+                if (this._prevTokenType === PhpLexer.SemiColon || this._prevTokenType === PhpLexer.Colon || this._prevTokenType === PhpLexer.OpenCurlyBracket || this._prevTokenType === PhpLexer.CloseCurlyBracket) {
+                    token = super.nextToken()
+                } else {
+                    token = new CommonToken(undefined, PhpLexer.SemiColon, undefined, undefined, undefined);
+                    token.text = ';';
+                }
+            }
+        }
+
+        else if (token.type === PhpLexer.HtmlName) {
+            this._htmlNameText = token.text
+        }
+
+        else if (token.type === PhpLexer.HtmlDoubleQuoteString) {
+            if (token.text === "php" && this._htmlNameText === "language") {
+                this._phpScript = true;
+            }
+        }
+
+        else if (this._mode === PhpLexer.HereDoc) {
+            // Heredoc and Nowdoc syntax support: http://php.net/manual/en/language.types.string.php#language.types.string.syntax.heredoc
+            if (token.type === PhpLexer.StartHereDoc || token.type === PhpLexer.StartNowDoc) {
+                this._heredocIdentifier = token.text.slice(3).trim().replace(/\'$/, '');
```

Comment:
This code does not correspond to the CSharp port. This fails because in C#, `trim(char)` removes both the leading and trailing `char`. But, in TypeScript, `replace(pat, string)` only replaces the first occurrence of the pattern. So, `'NOWDOC` is replaced with `NOWDOC'`. The way I did this is to have two replace(), one for the beginning of the string, and another for the end. Or, you could just call `replace(/\'/, '')` twice.